File: go/parsers/event.go
package parsers

import (
	"google.golang.org/protobuf/proto"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-go/serializers"
	"github.com/xconnio/wampproto-go/util"
	"github.com/xconnio/wampproto-protobuf/go/gen"
)

type Event struct {
	gen *gen.Event
	ex  *PayloadExpander
}

func NewEventFields(gen *gen.Event, payload []byte) messages.EventFields {
	return &Event{
		gen: gen,
		ex:  &PayloadExpander{payload: payload, serializer: gen.GetPayloadSerializerId()},
	}
}

func (e *Event) SubscriptionID() uint64 {
	return e.gen.SubscriptionId
}

func (e *Event) PublicationID() uint64 {
	return e.gen.PublicationId
}

func setDetail(details *map[string]any, key string, value any) {
	if *details == nil {
		*details = make(map[string]any)
	}

	(*details)[key] = value
}

func (e *Event) Details() map[string]any {
	var details map[string]any

	if e.gen.Publisher > 0 {
		setDetail(&details, "publisher", e.gen.Publisher)
	}

	if e.gen.PublisherAuthid != "" {
		setDetail(&details, "publisher_authid", e.gen.PublisherAuthid)
	}

	if e.gen.PublisherAuthrole != "" {
		setDetail(&details, "publisher_authrole", e.gen.PublisherAuthrole)
	}

	if e.gen.Topic != "" {
		setDetail(&details, "topic", e.gen.Topic)
	}

	return details
}

func (e *Event) Args() []any {
	return e.ex.Args()
}

func (e *Event) KwArgs() map[string]any {
	return e.ex.Kwargs()
}

func (e *Event) PayloadIsBinary() bool {
	return true
}

func (e *Event) Payload() []byte {
	return e.ex.Payload()
}

func (e *Event) PayloadSerializer() uint64 {
	return e.gen.GetPayloadSerializerId()
}

func EventToProtobuf(event *messages.Event) ([]byte, error) {
	msg := &gen.Event{
		SubscriptionId: event.SubscriptionID(),
		PublicationId:  event.PublicationID(),
	}

	if publisher, ok := util.AsUInt64(event.Details()["publisher"]); ok {
		msg.Publisher = publisher

		authID, ok := util.AsString(event.Details()["publisher_authid"])
		if ok {
			msg.PublisherAuthid = authID
		}

		authRole, ok := util.AsString(event.Details()["publisher_authrole"])
		if ok {
			msg.PublisherAuthrole = authRole
		}

		topic, ok := util.AsString(event.Details()["topic"])
		if ok {
			msg.Topic = topic
		}
	}

	payloadSerializer := selectPayloadSerializer(event.Details())
	msg.PayloadSerializerId = payloadSerializer
	payload, err := serializers.SerializePayload(payloadSerializer, event.Args(), event.KwArgs())
	if err != nil {
		return nil, err
	}

	data, err := proto.Marshal(msg)
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeEvent, data, payload), nil
}

func ProtobufToEvent(data []byte, payload []byte) (*messages.Event, error) {
	msg := &gen.Event{}
	err := proto.Unmarshal(data, msg)
	if err != nil {
		return nil, err
	}

	return messages.NewEventWithFields(NewEventFields(msg, payload)), nil
}

File: python/wampprotobuf/parsers/event.py
from typing import Any

from wampproto.messages import Message
from wampproto.messages.event import IEventFields, Event

from wampprotobuf.gen import event_pb2
from wampprotobuf.parsers import helpers


class EventFields(IEventFields):
    def __init__(self, msg: event_pb2.Event):
        super().__init__()
        self._msg = msg
        self._args = None
        self._kwargs = None
        self._unpacked = False

    @property
    def subscription_id(self) -> int:
        return self._msg.subscription_id

    @property
    def publication_id(self) -> int:
        return self._msg.publication_id

    def unpack(self):
        try:
            self._unpacked = True
            args, kwargs = helpers.from_cbor_payload(self._msg.payload)
            self._args = args
            self._kwargs = kwargs
        except Exception as e:
            print(f"error parsing CBOR payload: {e}")

    @property
    def args(self) -> list[Any] | None:
        if not self._unpacked:
            self.unpack()

        return self._args

    @property
    def kwargs(self) -> dict[str, Any] | None:
        if not self._unpacked:
            self.unpack()

        return self._kwargs

    @property
    def details(self) -> dict[str, Any]:
        return {}

    def payload_is_binary(self) -> bool:
        return True

    @property
    def payload(self) -> bytes | None:
        return self._msg.payload

    @property
    def payload_serializer(self) -> int:
        return self._msg.payload_serializer


def from_protobuf(payload: bytes) -> Message:
    result = event_pb2.Event()
    result.ParseFromString(payload)

    return Event(EventFields(result))


def to_protobuf(event: Event) -> bytes:
    result = event_pb2.Event()
    result.subscription_id = event.subscription_id
    result.publication_id = event.publication_id

    if event.payload_is_binary():
        result.payload = event.payload
        result.payload_serializer = event.payload_serializer
    else:
        try:
            payload = helpers.to_cbor_payload(event.args, event.kwargs)
        except Exception as e:
            raise Exception(f"error in serialization to cbor: {e}")

        serializer = 1

        result.payload = payload
        result.payload_serializer = serializer

    return result.SerializeToString()
